stablehlo/reference/Ops.cpp

```diff
@@ -133,6 +200,294 @@ Tensor evalConvertOp(const Tensor &operand, TensorType resultType) {
+          featureGroupCount, /*batchGroupCount=*/1,
+          RankedTensorType::get(inferredConvolutionType[0].getDims(),
+                                result.getElementType())
+              .cast<TensorType>());
+      results.push_back(convolutionResult);
+    }
+    return evalConcatenateOp(results, outputFeatureDimension, result.getType());
+  }
+  auto lhsShape = [&](auto n, auto hw, auto c) {
+    SmallVector<int64_t> permInput;
+    permInput.push_back(n);
+    permInput.append(hw.begin(), hw.end());
+    permInput.push_back(c);
+    SmallVector<int64_t> permDims;
+    permDims.push_back(inputBatchDimension);
+    permDims.append(inputSpatialDimensions.begin(),
+                    inputSpatialDimensions.end());
+    permDims.push_back(inputFeatureDimension);
+    SmallVector<int64_t> permOutput(permDims.size());
+    for (auto [idx, dim] : llvm::enumerate(permDims))
+      permOutput[dim] = permInput[idx];
+    return permOutput;
+  };
+  auto lhsShapePadding = [&](auto n, auto hw, auto c) {
+    SmallVector<SmallVector<int64_t>> permInput;
+    permInput.push_back(n);
+    permInput.append(hw.begin(), hw.end());
+    permInput.push_back(c);
+    SmallVector<int64_t> permDims;
+    permDims.push_back(inputBatchDimension);
+    permDims.append(inputSpatialDimensions.begin(),
+                    inputSpatialDimensions.end());
+    permDims.push_back(inputFeatureDimension);
+    SmallVector<SmallVector<int64_t>> permOutput(permDims.size());
+    for (auto [idx, dim] : llvm::enumerate(permDims))
+      permOutput[dim] = permInput[idx];
+    return permOutput;
+  };
+
+  auto lhsWindowDimensions =
+      lhsShape(lhs.getShape()[inputBatchDimension],
+               extractElements(rhs.getShape(), kernelSpatialDimensions),
+               lhs.getShape()[inputFeatureDimension]);
+  auto lhsWindowStrides = lhsShape(1, windowStrides, 1);
+  auto lhsPadding = lhsShapePadding(SmallVector<int64_t>(2, 0), padding,
+                                    SmallVector<int64_t>(2, 0));
+  auto lhsBaseDilations = lhsShape(0, Sizes(lhsDilation) - 1, 0);
+  auto lhsWindowDilations = lhsShape(1, rhsDilation, 1);
+
+  auto outputSpacialIndexIt = IndexSpaceIterator(
+      Sizes(extractElements(result.getShape(), outputSpatialDimensions)),
+      Index(outputSpatialDimensions.size()));
+
+  auto outputSpacialIndexItEnd = IndexSpaceIterator(
+      Sizes(extractElements(result.getShape(), outputSpatialDimensions)),
+      std::nullopt);
+  auto resultElementType = result.getElementType();
+  for (; outputSpacialIndexIt != outputSpacialIndexItEnd;
+       ++outputSpacialIndexIt) {
+    SmallVector<int64_t> lhsPaddingLow;
+    SmallVector<int64_t> lhsPaddingHigh;
+    for (auto paddingPair : lhsPadding) {
+      lhsPaddingLow.push_back(paddingPair[0]);
+      lhsPaddingHigh.push_back(paddingPair[1]);
+    }
+    SmallVector<Type> inferredPadType;
+    auto padStatus = hlo::inferPadOp(
```

Comment:
Common for all the inferOp* used:

The boilerplate needed for the inferOp calls and the call itself cause  distraction towards our goal of having a readable semantics. This is true even if we use a common interface for evalOp and inferOp. 
My proposal would be to encapsulate all the boilerplate code + the call to inferOp in a function like

`TensorType getPadOpType(Sizes paddingLow, Sizes paddingHigh, Sizes lhsBaseDilations);`

and use it as
```
auto paddedLhs = evalPadOp(lhs, getZeroScalarTensor(resultElementType),
                               Sizes(lhsPaddingLow), Sizes(lhsBaseDilations),
                               getPadOpType(...));
```

This is help improve the readability a lot.